File: packages/go/analytics.go
package opensms

import (
	"context"
	"net/http"
	"net/url"
)

// Analytics reads delivery and spend metrics, reached as client.Analytics.
type Analytics struct {
	http *transport
}

func analyticsQuery(p AnalyticsParams) url.Values {
	return newQuery().
		str("currency", p.Currency).
		str("range", p.Range).
		time("from", p.From).
		time("to", p.To).
		str("bucket", p.Bucket).
		values()
}

// Overview returns totals for the period (GET /v1/analytics/overview).
func (r *Analytics) Overview(ctx context.Context, params AnalyticsParams) (*AnalyticsOverview, error) {
	return call[AnalyticsOverview](ctx, r.http, request{method: http.MethodGet, path: "/v1/analytics/overview", query: analyticsQuery(params)})
}

// ByCountry breaks metrics down by country (GET /v1/analytics/by-country).
func (r *Analytics) ByCountry(ctx context.Context, params AnalyticsParams) ([]AnalyticsBreakdown, error) {
	return callSlice[AnalyticsBreakdown](ctx, r.http, request{method: http.MethodGet, path: "/v1/analytics/by-country", query: analyticsQuery(params)})
}

// ByCarrier breaks metrics down by carrier (GET /v1/analytics/by-carrier).
func (r *Analytics) ByCarrier(ctx context.Context, params AnalyticsParams) ([]AnalyticsBreakdown, error) {
	return callSlice[AnalyticsBreakdown](ctx, r.http, request{method: http.MethodGet, path: "/v1/analytics/by-carrier", query: analyticsQuery(params)})
}

// BySenderID breaks metrics down by sender ID
// (GET /v1/analytics/by-sender-id).
func (r *Analytics) BySenderID(ctx context.Context, params AnalyticsParams) ([]AnalyticsBreakdown, error) {
	return callSlice[AnalyticsBreakdown](ctx, r.http, request{method: http.MethodGet, path: "/v1/analytics/by-sender-id", query: analyticsQuery(params)})
}

// Timeseries returns metrics per bucket (GET /v1/analytics/timeseries).
func (r *Analytics) Timeseries(ctx context.Context, params AnalyticsParams) ([]AnalyticsPoint, error) {
	return callSlice[AnalyticsPoint](ctx, r.http, request{method: http.MethodGet, path: "/v1/analytics/timeseries", query: analyticsQuery(params)})
}

File: packages/java/src/main/java/io/opensms/Analytics.java
package io.opensms;

import io.opensms.models.AnalyticsBreakdown;
import io.opensms.models.AnalyticsOverview;
import io.opensms.models.AnalyticsPoint;

import java.util.List;

/** Delivery and spend analytics. Accessed as {@code client.analytics()}. */
public final class Analytics {

    private final ApiTransport t;

    Analytics(ApiTransport t) {
        this.t = t;
    }

    /** @return totals for the last 30 days. */
    public AnalyticsOverview overview() {
        return overview(new AnalyticsParams());
    }

    /**
     * @param q range, currency and bucket.
     * @return totals.
     */
    public AnalyticsOverview overview(AnalyticsParams q) {
        return t.get("/v1/analytics/overview", q.toQuery(), ApiTransport.type(AnalyticsOverview.class));
    }

    /** @return metrics per country for the last 30 days. */
    public List<AnalyticsBreakdown> byCountry() {
        return byCountry(new AnalyticsParams());
    }

    /**
     * @param q range, currency and bucket.
     * @return metrics per country.
     */
    public List<AnalyticsBreakdown> byCountry(AnalyticsParams q) {
        return t.get("/v1/analytics/by-country", q.toQuery(), ApiTransport.listOf(AnalyticsBreakdown.class));
    }

    /** @return metrics per carrier for the last 30 days. */
    public List<AnalyticsBreakdown> byCarrier() {
        return byCarrier(new AnalyticsParams());
    }

    /**
     * @param q range, currency and bucket.
     * @return metrics per carrier.
     */
    public List<AnalyticsBreakdown> byCarrier(AnalyticsParams q) {
        return t.get("/v1/analytics/by-carrier", q.toQuery(), ApiTransport.listOf(AnalyticsBreakdown.class));
    }

    /** @return metrics per sender ID for the last 30 days. */
    public List<AnalyticsBreakdown> bySenderId() {
        return bySenderId(new AnalyticsParams());
    }

    /**
     * @param q range, currency and bucket.
     * @return metrics per sender ID.
     */
    public List<AnalyticsBreakdown> bySenderId(AnalyticsParams q) {
        return t.get("/v1/analytics/by-sender-id", q.toQuery(), ApiTransport.listOf(AnalyticsBreakdown.class));
    }

    /** @return metrics per day for the last 30 days. */
    public List<AnalyticsPoint> timeseries() {
        return timeseries(new AnalyticsParams());
    }

    /**
     * @param q range, currency and bucket.
     * @return metrics per bucket.
     */
    public List<AnalyticsPoint> timeseries(AnalyticsParams q) {
        return t.get("/v1/analytics/timeseries", q.toQuery(), ApiTransport.listOf(AnalyticsPoint.class));
    }
}
